### PR TITLE
Add ConfigRecognizer API

### PR DESCRIPTION
**Pull request by blazember**
This commit introduces the `ConfigRecognition` API that is meant to
recognize if a provided declarative configuration is recognized by the
rules defined in a given implementation. The main use case for this
implementation is to recognize member, client and failover client XML
and YAML configurations just by looking into the content of the
configuration, without building any actual configuration.

Along with the API the following three implementations are added:
- `MemberConfigRecognizer` for recognizing member XML and YAML
configurations
- `ClientConfigRecognizer` for recognizing client XML and YAML
configurations
- `ClientFailoverConfigRecognizer` for recognizing failover client XML and
YAML configurations

All the three are extensible with custom recognizers so that even those
configurations can be recognized that would otherwise remain
unrecognized with the built-in set of recognizers that recognizes XML
and YAML configurations by checking the root node in the provided
configuration.

The API recognizes configuration provided in InputStreams only, while
the ConfigBuilder implementations can be built from files defined by
their location and URLs too. The reason for lacking the support for
these two options is that both are easy to convert to InputStreams and
that handling errors caused by non-existing resources, missing
privileges etc should be handled outside of this API with potentially
having more contextual information.

The built-in implementations honor parse errors with unrecognized
configuration. The reason is that the provided InputStream can be tested
both for XML and YAML configuration and at least one is expected to
lead to parse errors.

Implements #16866

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/563/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/563/Hazelcast-pr-builder-563.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/AbstractXmlConfigRootTagRecognizer.java:73:25: Variable 'isMemberXml' explicitly initialized to 'false' (default value for its type). [ExplicitInitialization]
--------------------------
</details>
</pre>

**Comment by blazember** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigStreamRecognizer.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.config;
+
+public interface ConfigStreamRecognizer {
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ConfigStreamRecognizer.java:19]: Leftover.

File: hazelcast/src/main/java/com/hazelcast/config/ConfigStream.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * If calling {@code reset()} on the provided implementation fails with
+ * an exception, this class reads the stream into a {@code byte[]} and
+ * delegates all {@link InputStream} method calls to a
+ * {@link ByteArrayInputStream} created with this {@code byte[]}. To
+ * prevent OOM issues, the size of this {@code byte[]} is limited to 4096
+ * bytes. This limit can be configured in the constructor.
+ *
+ * @see ConfigRecognizer
+ */
+public class ConfigStream extends InputStream {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/config/ConfigStream.java:43]: Minor: does this class need to be public API?

File: hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigRecognizer.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public AbstractConfigRecognizer(Collection<ConfigRecognizer> recognizers) {
+        this.recognizers = recognizers;
+    }
+
+    @Override
+    public boolean isRecognized(ConfigStream configStream) throws Exception {
+        boolean recognized = false;
+        for (ConfigRecognizer recognizer : recognizers) {
+            configStream.reset();
+            recognized = recognized || recognizer.isRecognized(configStream);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigRecognizer.java:40]: Minor: can we exit early if it was recognized?

File: hazelcast/src/main/java/com/hazelcast/config/ConfigStream.java (hunk 2 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/config/ConfigStream.java:43]: Actually found out it needs to, nevermind.

**Pull request merged**